Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -57,8 +56,18 @@ void WebPermissionController::query(WebCore::ClientOrigin&& origin, WebCore::Per
         proxyIdentifier = WebPage::fromCorePage(*page).webPageProxyIdentifier();
     }
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), proxyIdentifier, source, WTFMove(completionHandler) });
-    tryProcessingRequests();
+    if (async == WebCore::Asynchronous::Yes) {
+        m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), proxyIdentifier, source, WTFMove(completionHandler) });
+        tryProcessingRequests();
+    } else {
+        std::optional<WebCore::PermissionState> state;
+        bool shouldCache;
+
+        WebProcess::singleton().sendSync(Messages::WebPermissionControllerProxy::QuerySync(origin, descriptor, proxyIdentifier, source), Messages::WebPermissionControllerProxy::QuerySync::Reply(state, shouldCache));
+        completionHandler(state);
+        if (shouldCache && state)
+            updateCache(origin, descriptor, *state);
```

Comment:
I agree, updateCache cache should come before the completion handler as well. I'll change this in `WebPermissionController::tryProcessingRequests` as well. But I feel that we can do this and avoid the issue without adding a new function since that would duplicate code no?

Reply:
The querySync is to make it clear for the caller that the function will return sync-ly; we usually don't let caller pass CompletionHandler for pure sync operation (the WebCore::Synchronous mode).
For the duplicate code, we can add helper functions to share code between two functions, but I don't think the duplicate part is that much (query cache and get identifier). Given that the cache is not actually in use and may get removed, it may be even less.